src/stream/src/executor/source.rs

```diff
@@ -212,24 +214,67 @@ impl SourceReader {
     }
 }
 
+macro_rules! impl_take_snapshot {
+    ([], $state_store: expr, $epoch: ident, $vec_split_impl: ident, $({$connector: ident, $connector_split_type: ident}), *) => {
+        paste! {
+            match $vec_split_impl[0].get_type().as_str() {
+                $(
+                    $connector_split_type => {
+                        let type_cache = $vec_split_impl.iter().map(|split_impl| split_impl.[<get_ $connector>]()).collect::<anyhow::Result<Vec<_>>>().to_rw_result()?;
+                        $state_store.take_snapshot(type_cache, $epoch).await.to_rw_result()?;
+                    },
+                )*
+                _ => todo!(),
+            }
+        }
+    };
+}
+
 impl<S: StateStore> SourceExecutor<S> {
+    async fn take_snapshot(&mut self, epoch: u64) -> Result<()> {
+        let cache = self
+            .state_cache
+            .iter()
+            .map(|(_, split_impl)| split_impl)
+            .collect::<Vec<_>>();
+        if !cache.is_empty() {
+            impl_take_snapshot!([], self.split_state_store, epoch, cache,
+                { kafka, KAFKA_SPLIT_TYPE },
+                { kinesis, KINESIS_SPLIT_TYPE },
+                { nexmark, NEXMARK_SPLIT_TYPE },
+                { pulsar, PULSAR_SPLIT_TYPE }
+            );
```

Comment:
@shanicky we should not manually list connectors here. any solution?

Reply:
It should be possible to ensure by more trait restrictions, directly through the `build.rs` to scan the directory connector to generate code directly, but I have not tried to know if it is possible, and then xx_split_type has gone, do not forget to rebase a main

Reply:
let's leave it for now